xdsl/dialects/pdl.py

```diff
@@ -0,0 +1,322 @@
+                arg_typ = cast(RangeType[AnyPDLType], arg.typ)
+                return arg_typ.element_type
+            else:
+                return arg.typ
+
+        if len(self.arguments) > 0:
+            elem_type = get_type_or_elem_type(self.result)
+
+            for arg in self.arguments:
+                if cur_elem_type := get_type_or_elem_type(arg) != elem_type:
+                    raise VerifyException(
+                        f"All arguments must have the same type or be an array  \
+                          of the corresponding element type. First element type:\
+                          {elem_type}, current element type: {cur_elem_type}")
+
+
+@irdl_op_definition
+class ReplaceOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlreplace-mlirpdlreplaceop
+    """
+    name: str = "pdl.replace"
+    opValue: Annotated[Operand, OperationType]
+    replOperation: Annotated[OptOperand, OperationType]
+    replValues: Annotated[VarOperand, ValueType | ArrayAttr[ValueType]]
+
+    irdl_options = [AttrSizedOperandSegments()]
+
+
+@irdl_op_definition
+class ResultOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlresult-mlirpdlresultop
+    """
+    name: str = "pdl.result"
+    index: OpAttr[IntegerAttr[IntegerType]]
+    parent_: Annotated[Operand, OperationType]
+    val: Annotated[OpResult, ValueType]
+
+
+@irdl_op_definition
+class ResultsOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlresults-mlirpdlresultsop
+    """
+    name: str = "pdl.results"
+    index: OpAttr[IntegerAttr[IntegerType]]
+    parent_: Annotated[Operand, OperationType]
+    val: Annotated[OpResult, ValueType | ArrayAttr[ValueType]]
+
+
+@irdl_op_definition
+class RewriteOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlrewrite-mlirpdlrewriteop
+    """
+    name: str = "pdl.rewrite"
+    root: Annotated[OptOperand, OperationType]
+    # name of external rewriter function
+    # https://github.com/xdslproject/xdsl/issues/98
+    # name: OptOpAttr[StringAttr]
+    externalArgs: Annotated[VarOperand, AnyPDLType]
+    body: OptRegion
+
+    irdl_options = [AttrSizedOperandSegments()]
+
+    def verify_(self) -> None:
+        if 'name' in self.attributes:
+            if not isinstance(self.attributes['name'], StringAttr):
+                raise Exception("expected 'name' attribute to be a StringAttr")
+
+    @staticmethod
+    def get(name: str, root: SSAValue | None, external_args: list[SSAValue],
+            body: Region | None) -> RewriteOp:
+
+        operands: list[SSAValue | list[SSAValue]] = []
+        if root is not None:
+            operands.append([root])
+        else:
+            operands.append([])
+        operands.append(external_args)
+
+        regions: list[Region | list[Region]] = []
+        if body is not None:
+            regions.append(body)
+        else:
+            regions.append([])
```

Comment:
Does this actually end up giving you no region in the right cases? I am not sure anymore how we coded this, but once upon a time, [[]] for the regions argument would be translated to an empty region (AFAIR).

Reply:
Yeah, and it's something that I'd like to fix up in follow-up work. `irdl_op_builder` claims that it can take `None` as an element of the argument sequence, but it doesn't actually handle it, and translates an empty list to the right value instead.